cmd/snap/cmd_run.go

```diff
@@ -1156,55 +1164,88 @@ var osGetpid = os.Getpid
 
 var cgroupProcessPathInTrackingCgroup = cgroup.ProcessPathInTrackingCgroup
 
+var errCannotTrackProcess = errors.New("cannot track application process")
+
 var createTransientScope = func(securityTag string) error {
 	if !features.RefreshAppAwareness.IsEnabled() {
 		return nil
 	}
 	logger.Debugf("creating transient scope %s", securityTag)
-	// The scope is created with a DBus call to systemd running either on
-	// system or session bus, depending on if we are starting a program as root
-	// or as a regular user.
-	var conn *dbus.Conn
-	var err error
-	if uid := osGetuid(); uid == 0 {
-		conn, err = dbusSystemBus()
-	} else {
-		conn, err = dbusSessionBus()
-		// XXX: Ideally we would check for a distinct error type but this is
-		// just an errors.New() in go-dbus code. To avoid being fragile ignore
-		// all errors when establishing session bus connection to avoid
-		// breaking user interactions. This is consistent with similar failure
-		// modes below, where other parts of the stack fail.
-		if err != nil {
-			logger.Debugf("user %d does not have a working DBus session bus", uid)
-			logger.Debugf("snapd cannot track the started application")
-			logger.Debugf("snap refreshes will not be postponed by this process")
-			return nil
-		}
-	}
-	if err != nil {
-		return err
-	}
+
 	// We ask the kernel for a random UUID. We need one because each transient
 	// scope needs a unique name. The unique name is comprosed of said UUID and
 	// the snap security tag.
 	uuid, err := randomUUID()
 	if err != nil {
 		return err
 	}
+
 	// Enforcing uniqueness is preferred to reusing an existing scope for
 	// simplicity since doing otherwise and joining an existing scope has
 	// limitations:
 	// - the originally started scope must be marked as a delegate, with all consequences.
 	// - the method AttachProcessesToUnit is unavailable on Ubuntu 16.04
+	//
 	// TODO: invert the arguments so that security tag comes first, UUID
 	// second. This will result in nicer scope names but needs to be
 	// coordinated with code in sandbox/cgroup and with the spread tests
 	// exercising that.
 	unitName := fmt.Sprintf("snap.%s.%s.scope", uuid, strings.TrimPrefix(securityTag, "snap."))
 	pid := osGetpid()
+
+	// The scope is created with a DBus call to systemd running either on
+	// system or session bus. We have a preference for session bus, as this is
+	// where applications normally go to. When a session bus is not available
+	// and the invoking user is root, we use the system bus instead.
+	//
+	// It is worth noting that hooks will not normally have a session bus to
+	// connect to, as they are invoked as descendants of snapd, and snapd is a
+	// service running outside of any session.
+	isSessionBus := true
+	conn, err := dbusSessionBus()
```

Comment:
Do we know if gdobus doesn't try to call dbus-launch?

Reply:
It does though I have not seen it succeed. I'll check if that was only available in the more recent version I used or in the version we locked via vendor.json

Reply:
I have now worked around go-dbus' annoying launching property.